File: 401-code-challenges/challenges/blogs/merge_sort/merge_sort.js
'use strict';

const arr = [8,4,23,42,16,15];

function sort(arr) {
  arr.sort((a, b) => a-b);
  return arr;
}

function mergeSort (arr) {
  let n = arr.length;
  let left = [];
  let right = [];
  let mid = n/2;

  if (n > 1) {

    for (let i = 0; i < mid; i++ ) {
      
      left.push(arr[i])
    }
    for (let i = mid; i < arr.length; i++)  {

      right.push(arr[i])
    }

  }

  return merge(sort(left), sort(right))

}

function merge(left, right)  {
  const arr = [];
  let i = 0;
  let j = 0;
  let k = 0;

  while(i < left.length || j < right.length)  {

    if(left[i] <= right[j]) {
      arr[k] = left[i]
      i = i+1
    } else {
      arr[k] = right[j]
      j = j+1
    }

    k = k+1

  }

  console.log(arr)
  return arr;
}

mergeSort(arr)

module.exports = {
  sort,
  merge,
  mergeSort,
}

File: 401-code-challenges/javaChallenges/src/main/java/javaChallenges/MergeSort.java
package javaChallenges;

public class MergeSort {

    public void mergeSort(int[] arr) {
        int n = arr.length;
        if(n > 1) {
        
            int mid = n/2;
            int[] left = new int[mid];
            int[] right = new int[n - mid];

            for(int i = 0; i < left.length; i++) {
                left[i] = arr[i];
            }
            for(int j = 0; j < right.length; j++) {
                right[j] = arr[j + mid];
            }

            mergeSort(left);
            mergeSort(right);
            merge(left, right, arr);
        }

    }

    public void merge(int [] left, int [] right, int [] arr) {

        int i = 0;
        int j = 0;
        int k = 0;

        while (i < left.length && j < right.length) {
            if(left[i] <= right[j]) {
                arr[k] = left[i];
                i++;
                k++;
            } else {
                arr[k] = right[j];
                j++;
                k++;
            }
            k++;
        }

        if(i == left.length) {
            for(int l = j; l < right.length; l++) {
                arr[k] = right[l];
            }
        } else {
            for(int m = i; m < left.length; m++) {
                arr[k] = left[m];
            }
        }
    }
}
